File: CPP14 GCC 5.4.1/ABC001/ABC001_C_v01.cpp
/* AtCoder Beginner Contest 001 C */
/*

[URL]
https://atcoder.jp/contests/abc001/tasks/abc001_3

[compiler]
C++14 (GCC 5.4.1)

[reference]
<1> https://cpprefjp.github.io/reference/cmath/round.html

*/

#include <cmath>
#include <iostream>
#include <string>

// 風程(wind run)から風力(wind scale)を算出する
unsigned int calc_wind_scale( unsigned int wind_run )
{
	/* local variables for this function */
	unsigned int wind_speed;

	// STEP.01
	// convert the `wind run` to `wind speed`
	wind_speed = std::round( ((double)wind_run)/6.0e+00 );

	// STEP.02
	// determine the `wind scale` to output
	if      ( wind_speed <=   2 ) { return  0; }
	else if ( wind_speed <=  15 ) { return  1; }
	else if ( wind_speed <=  33 ) { return  2; }
	else if ( wind_speed <=  54 ) { return  3; }
	else if ( wind_speed <=  79 ) { return  4; }
	else if ( wind_speed <= 107 ) { return  5; }
	else if ( wind_speed <= 138 ) { return  6; }
	else if ( wind_speed <= 171 ) { return  7; }
	else if ( wind_speed <= 207 ) { return  8; }
	else if ( wind_speed <= 244 ) { return  9; }
	else if ( wind_speed <= 284 ) { return 10; }
	else if ( wind_speed <= 326 ) { return 11; }
	else                          { return 12; }
}

// 風力(wind scale)を考慮した風向（文字列）を返す
std::string determine_wind_direction( unsigned int wind_scale, unsigned int wind_direction_num )
{
	/* local variables for this function */
	unsigned int wind_direction_num_scaled;

	// STEP.01
	// scale the wind direction
	wind_direction_num_scaled = wind_direction_num * 10;

	// STEP.02
	// determine the `wind direction` to output
	if ( wind_scale == 0 ) { return "C"; }

	if ( wind_direction_num_scaled < 1125 ) { return "N"; }
	else
	{
		if      ( wind_direction_num_scaled <  3375 ) { return "NNE"; }
		else if ( wind_direction_num_scaled <  5625 ) { return "NE";  }
		else if ( wind_direction_num_scaled <  7875 ) { return "ENE"; }
		else if ( wind_direction_num_scaled < 10125 ) { return "E";   }
		else if ( wind_direction_num_scaled < 12375 ) { return "ESE"; }
		else if ( wind_direction_num_scaled < 14625 ) { return "SE";  }
		else if ( wind_direction_num_scaled < 16875 ) { return "SSE"; }
		else if ( wind_direction_num_scaled < 19125 ) { return "S";   }
		else if ( wind_direction_num_scaled < 21375 ) { return "SSW"; }
		else if ( wind_direction_num_scaled < 23625 ) { return "SW";  }
		else if ( wind_direction_num_scaled < 25875 ) { return "WSW"; }
		else if ( wind_direction_num_scaled < 28125 ) { return "W";   }
		else if ( wind_direction_num_scaled < 30375 ) { return "WNW"; }
		else if ( wind_direction_num_scaled < 32625 ) { return "NW";  }
		else if ( wind_direction_num_scaled < 34875 ) { return "NNW"; }
		else                                          { return "N";   }
	}
}

int main (void)
{
	/* constants for main process */
	
	/* variables for main process */
	char     wind_direction_chr;
	unsigned int wind_direction_num;
	unsigned int wind_run;
	unsigned int wind_scale;


	// STEP.01
	// read out the given data
	std::cin >> wind_direction_num; // 風向
	std::cin >> wind_run;           // 風程

	// STEP.02
	// calculate the `wind scale`
	wind_scale = calc_wind_scale( wind_run );

	// STEP.03
	// output the result
	std::cout << determine_wind_direction( wind_scale, wind_direction_num ) << " " << wind_scale << std::endl;

	// STEP.END
	return 0;
}

File: Haskell/ABC/ABC001/ABC001_C_v01.hs
-- [contest]  ABC 001
-- [task]     C
-- [URL]      https://atcoder.jp/contests/abc001/tasks/abc001_3
-- [compiler] Haskell (GHC 7.10.3)
-- [status]   NOT Submitted

-- modules to import --

import qualified Data.ByteString.Char8
import qualified Data.Maybe
import qualified Text.Printf
import           Prelude

-- functions for this task --

tuplify2 (x:y:_) = (x,y)
tuplify2 _       = Prelude.undefined

read_Integer      = Prelude.fst . Data.Maybe.fromJust . Data.ByteString.Char8.readInteger
read_TupleInteger = tuplify2 . Prelude.map read_Integer . Data.ByteString.Char8.words

get_Integer      = read_Integer      <$> Data.ByteString.Char8.getLine
get_TupleInteger = read_TupleInteger <$> Data.ByteString.Char8.getLine

wind_Direction :: Integer -> Integer -> Data.ByteString.Char8.ByteString
wind_Direction degree wind_force
        | wind_force == 0  = Data.ByteString.Char8.pack $ "C"
        |Prelude.otherwise = wind_Direction_core $ degree
        where
                wind_Direction_core :: Integer -> Data.ByteString.Char8.ByteString
                wind_Direction_core degree
                        | degree_scaled >=  11.25 && degree_scaled <=  33.74 = Data.ByteString.Char8.pack $ "NNE"
                        | degree_scaled >=  33.75 && degree_scaled <=  56.24 = Data.ByteString.Char8.pack $ "NE"
                        | degree_scaled >=  56.25 && degree_scaled <=  78.74 = Data.ByteString.Char8.pack $ "ENE"
                        | degree_scaled >=  78.75 && degree_scaled <= 101.24 = Data.ByteString.Char8.pack $ "E"
                        | degree_scaled >= 101.25 && degree_scaled <= 123.74 = Data.ByteString.Char8.pack $ "ESE"
                        | degree_scaled >= 123.75 && degree_scaled <= 146.24 = Data.ByteString.Char8.pack $ "SE"
                        | degree_scaled >= 146.25 && degree_scaled <= 168.74 = Data.ByteString.Char8.pack $ "SSE"
                        | degree_scaled >= 168.75 && degree_scaled <= 191.24 = Data.ByteString.Char8.pack $ "S"
                        | degree_scaled >= 191.25 && degree_scaled <= 213.74 = Data.ByteString.Char8.pack $ "SSW"
                        | degree_scaled >= 213.75 && degree_scaled <= 236.24 = Data.ByteString.Char8.pack $ "SW"
                        | degree_scaled >= 236.25 && degree_scaled <= 258.74 = Data.ByteString.Char8.pack $ "WSW"
                        | degree_scaled >= 258.75 && degree_scaled <= 281.24 = Data.ByteString.Char8.pack $ "W"
                        | degree_scaled >= 281.25 && degree_scaled <= 303.74 = Data.ByteString.Char8.pack $ "WNW"
                        | degree_scaled >= 303.75 && degree_scaled <= 326.24 = Data.ByteString.Char8.pack $ "NW"
                        | degree_scaled >= 326.25 && degree_scaled <= 348.74 = Data.ByteString.Char8.pack $ "NNW"
                        | Prelude.otherwise                                  = Data.ByteString.Char8.pack $ "N"
                        where
                                degree_scaled = (realToFrac $ degree) / (10.0 :: Double)

wind_force :: Integer -> Integer
wind_force wind_run
        | speed <=   2 =  0
        | speed <=  15 =  1
        | speed <=  33 =  2
        | speed <=  54 =  3
        | speed <=  79 =  4
        | speed <= 107 =  5
        | speed <= 138 =  6
        | speed <= 171 =  7
        | speed <= 207 =  8
        | speed <= 244 =  9
        | speed <= 284 = 10
        | speed <= 326 = 11
        where
                speed = Prelude.truncate $ (realToFrac $ wind_run) / (6.0 :: Double) + (0.5 :: Double)

-- the main process is as follows --

main :: IO ()
main = do

        -- STEP.01
        -- read out the given data
        given_data <- get_TupleInteger

        -- STEP.02
        -- calculate & output the answer of this task
        Prelude.print $                  Prelude.fst $ given_data
        Prelude.print $ wind_Direction $ Prelude.fst $ given_data
        Prelude.print $ Prelude.snd $ given_data

-- EOF --
